src/devices/SoftwareSpi/SoftwareSpi.cs

```diff
@@ -0,0 +1,199 @@
+
+            // TODO: To respect ClockFrequency we need to inject the right delays here
+            //       and have some very accurate way to measure time.
+            //       Ideally we should verify the output with an oscilloscope.
+            if (cpha)
+            {
+                _bitTransfer = new ScopeData(
+                    enter: () => {
+                        _controller.Write(_clk, !idle);
+                    },
+                    exit: () => {
+                        controller.Write(_clk, idle);
+                    });
+            }
+            else
+            {
+                _bitTransfer = new ScopeData(
+                    exit: () => {
+                        _controller.Write(_clk, !idle);
+                        _controller.Write(_clk, idle);
+                    });
+            }
+
+            _chipSelect = new ScopeData(
+                enter: () => {
+                    _controller.Write(_cs, _settings.ChipSelectLineActiveState);
+                },
+                exit: () => {
+                    _controller.Write(_cs, !(bool)_settings.ChipSelectLineActiveState);
+                });
+        }
+
+        public override SpiConnectionSettings ConnectionSettings => _settings;
+
+        public override void TransferFullDuplex(ReadOnlySpan<byte> dataToWrite, Span<byte> dataToRead)
+        {
+            if (dataToRead.Length != dataToRead.Length)
+            {
+                throw new ArgumentException(nameof(dataToWrite));
+            }
+
+            int bitLen = _settings.DataBitLength;
+            int lastBit = bitLen - 1;
+
+            using (ChipSelect)
+            {
+                for (int i = 0; i < dataToRead.Length; i++)
+                {
+                    byte readByte = 0;
+                    for (int j = 0; j < bitLen; j++)
+                    {
+                        using (BitTransfer)
+                        {
+                            int bit = _settings.DataFlow == DataFlow.MsbFirst ? lastBit - j : j;
+                            bool bitToWrite = ((dataToWrite[i] >> bit) & 1) == 1;
+                            if (ReadWriteBit(bitToWrite))
+                            {
+                                readByte |= (byte)(1 << bit);
+                            }
+                        }
+                    }
+
+                    dataToRead[i] = readByte;
+                }
+            }
+        }
+
+        private bool ReadWriteBit(bool bitToWrite)
+        {
+            _controller.Write(_mosi, bitToWrite);
+            return (bool)_controller.Read(_miso);
+        }
+
+        public override void Read(Span<byte> data)
+        {
+            Span<byte> dataToWrite = stackalloc byte[data.Length];
+            dataToWrite.Fill(0);
+            TransferFullDuplex(dataToWrite, data);
+        }
+
+        public override void Write(ReadOnlySpan<byte> data)
+        {
+            Span<byte> dataToRead = stackalloc byte[data.Length];
```

Comment:
If you only want to write, then you don't need this extra allocation, why not just send null and handle it in your TrasnferFullDuplex method?

Reply:
you still should be writing zeros when you read and this is almost no op